registry/quilt_server/auth.py

```diff
@@ -0,0 +1,527 @@
+    )
+
+ACTIVATE_SALT = 'activate'
+PASSWORD_RESET_SALT = 'reset'
+MAX_LINK_AGE = 60 * 60 * 24 # 24 hours
+
+def generate_activation_token(user_id):
+    existing_token = (
+        db.session.query(
+            ActivationToken
+        )
+        .filter(ActivationToken.user_id == user_id)
+        .one_or_none()
+    )
+    new_token = existing_token or ActivationToken(user_id=user_id, token=generate_uuid())
+    db.session.add(new_token)
+    db.session.commit()
+    return new_token.token
+
+def consume_activation_token(user_id, token):
+    token = (
+        db.session.query(
+            ActivationToken
+        ).filter(ActivationToken.user_id == user_id)
+        .filter(ActivationToken.token == token)
+        .one_or_none()
+    )
+    if not token:
+        return False
+    db.session.delete(token)
+    db.session.commit()
+    return True
+
+def generate_reset_token(user_id):
+    existing_token = (
+        db.session.query(
+            PasswordResetToken
+        ).filter(PasswordResetToken.user_id == user_id)
+        .one_or_none()
+    )
+    reset_token = existing_token or PasswordResetToken(user_id=user_id, token=generate_uuid())
+    db.session.add(reset_token)
+    db.session.commit()
+    return reset_token.token
+
+def consume_reset_token(user_id, token):
+    token = (
+        db.session.query(
+            PasswordResetToken
+        ).filter(PasswordResetToken.user_id == user_id)
+        .filter(PasswordResetToken.token == token)
+        .one_or_none()
+    )
+    if not token:
+        return False
+    db.session.delete(token)
+    db.session.commit()
+    return True
+
+
+def generate_activation_link(user_id):
+    token = generate_activation_token(user_id)
+    payload = {'id': user_id, 'token': token}
+    return linkgenerator.dumps(payload, salt=ACTIVATE_SALT)
+
+def generate_reset_link(user_id):
+    token = generate_reset_token(user_id)
+    payload = {'id': user_id, 'token': token}
+    return linkgenerator.dumps(payload, salt=PASSWORD_RESET_SALT)
+
+def verify_activation_link(link, max_age=None):
+    max_age = max_age if max_age is not None else MAX_LINK_AGE
+    try:
+        payload = linkgenerator.loads(link, max_age=max_age, salt=ACTIVATE_SALT)
+        if not consume_activation_token(payload['id'], payload['token']):
+            return None
+        return payload
+    except (TypeError, KeyError, ValueError, itsdangerous.BadData):
+        return None
+
+def verify_reset_link(link, max_age=None):
+    max_age = max_age if max_age is not None else MAX_LINK_AGE
+    try:
+        payload = linkgenerator.loads(link, max_age=max_age, salt=PASSWORD_RESET_SALT)
+        if not consume_reset_token(payload['id'], payload['token']):
+            return None
+        return payload
+    except (TypeError, KeyError, ValueError, itsdangerous.BadData):
+        return None
+
+def reset_password(user):
+    if not user:
+        return False
+    # set_unusable_password(user.name) # is this necessary?
```

Comment:
Pretty sure it is; it's the admin password reset feature, so need to disable the old password immediately.